earthdata/store.py

```diff
@@ -2,132 +2,266 @@
-            return SessionWithHeaderRedirection(
+            self._authenticated_req_session = req
+        client_kwargs = {
+            "auth": aiohttp.BasicAuth(
                 self.auth._credentials[0], self.auth._credentials[1]
-            )
+            ),
+            "trust_env": True,
+            "cookies": req.cookies.get_dict(),
+        }
+        if bearer_token and self.auth.authenticated:
+            client_kwargs["headers"] = {
+                "Authorization": f'Bearer {self.auth.token["access_token"]}'
+            }
+        session = fsspec.filesystem("https", client_kwargs=client_kwargs)
+        return session
 
-    def open(self, granules: List[Any], provider: str = None) -> List[Any]:
+    @singledispatchmethod
+    def open(
+        self,
+        granules: Union[List[str], List[DataGranule]],
+        provider: str = None,
+    ) -> List[Any]:
         """
-        returns a list of S3 file objects that can be used to access files
-        hosted on S3 by third party libraries like xarray
+        returns a list of fsspec file-like objects that can be used to access files
+        hosted on S3 or HTTPS by third party libraries like xarray.
+
         :param granules: a list of granules(DataGranule) instances or list of files
-        :param provider: cloud provider code, used when the list is made of s3 links
         :returns: a list of s3fs "file pointers" to s3 files
         """
+        print("granules need to be a list of URLs or a list of DataGranule instances")
+        return []
+
+    @open.register
+    def _open_granules(
+        self,
+        granules: List[DataGranule],
+        provider: str = None,
+    ) -> List[Any]:
+
         fileset: List = []
         data_links: List = []
-        if isinstance(granules[0], DataGranule):
-            provider = granules[0]["meta"]["provider-id"]
-            data_links = list(
-                chain.from_iterable(
-                    granule.data_links(access="direct") for granule in granules
-                )
+        if self.running_in_aws:
+            access_method = "direct"
+        else:
+            access_method = "on_prem"
+
+        provider = granules[0]["meta"]["provider-id"]
+        data_links = list(
+            chain.from_iterable(
+                granule.data_links(access=access_method) for granule in granules
             )
-            total_size = round(sum([granule.size() for granule in granules]) / 1024, 2)
+        )
+        total_size = round(sum([granule.size() for granule in granules]) / 1024, 2)
+        print(f" Opening {len(granules)} granules, approx size: {total_size} GB")
+
+        if self.auth is None:
             print(
-                f" Getting {len(granules)} granules, approx download size: {total_size} GB"
+                "A valid Earthdata login instance is required to retrieve credentials"
             )
-        elif isinstance(granules[0], str) and granules[0].startswith("s3"):
+            return []
```

Comment:
Oh, looks like in general you're not raising exceptions and returning empty for bad paths. I would assume some times an empty list is a valid return, so `None` or throwing an exception might make distinguishing the two easier.

Reply:
Changed to return `None` instead.